faux-mgs/src/main.rs

```diff
@@ -1387,6 +1436,96 @@ async fn run_command(
     }
 }
 
+async fn monorail_unlock(
+    log: &Logger,
+    sp: &SingleSp,
+    time_sec: u32,
+    key: Option<PathBuf>,
+) -> Result<()> {
+    let r = sp
+        .component_action_with_response(
+            SpComponent::MONORAIL,
+            ComponentAction::Monorail(
+                MonorailComponentAction::RequestChallenge,
+            ),
+        )
+        .await?;
+
+    let ComponentActionResponse::Monorail(
+        MonorailComponentActionResponse::RequestChallenge(challenge),
+    ) = r
+    else {
+        bail!("unexpected response: {r:?}");
+    };
+    info!(log, "received challenge {challenge:?}");
+
+    let response = match challenge {
+        UnlockChallenge::EcdsaSha2Nistp256(data) => {
+            let Some(priv_path) = key else {
+                bail!("need --key for ECDSA challenge");
+            };
+
+            debug!(log, "loading private key from {priv_path:?}");
+            let priv_key = ssh_key::PrivateKey::read_openssh_file(&priv_path)
+                .with_context(|| {
+                "failed to read private key at {priv_path}"
+            })?;
+            if priv_key.algorithm()
+                != (ssh_key::Algorithm::Ecdsa {
+                    curve: ssh_key::EcdsaCurve::NistP256,
+                })
+            {
+                bail!("invalid algorithm {:?}", priv_key.algorithm());
+            }
+
+            let signed = priv_key.sign(
+                "monorail-unlock",
+                ssh_key::HashAlg::Sha256,
+                &data,
+            )?;
+            debug!(log, "got signature {signed:?}");
+
+            let key_bytes =
+                signed.public_key().ecdsa().unwrap().as_sec1_bytes();
+            assert_eq!(key_bytes.len(), 65, "invalid key length");
+            let mut key = [0u8; 65];
+            key.copy_from_slice(key_bytes);
+
+            // Signature bytes are encoded per
+            // https://datatracker.ietf.org/doc/html/rfc5656#section-3.1.2
+            //
+            // They are a pair of `mpint` values, per
+            // https://datatracker.ietf.org/doc/html/rfc4251
+            //
+            // Each one is either 32 bytes or 33 bytes with a leading zero, so
+            // we'll awkwardly allow for both cases.
+            let sig_bytes = signed.signature_bytes();
+            let mut signature = [0u8; 64];
+            let start = match sig_bytes[3] {
+                32 => 4,
+                33 => 5,
+                i => bail!("invalid length {i}"),
+            };
+            signature[0..32].copy_from_slice(&sig_bytes[start..][..32]);
+            let tail = sig_bytes.len() - 32;
```

Comment:
Do we need to check the length of `sig_bytes` to ensure that this doesn't overlap with `[start..][..32]`?